### PR TITLE
Many improvements

### PR DESCRIPTION
**Pull request by surtich**
## Summary:

* Implemented Logout in client.
* Using media breakpoints for responsive behavior.
* Configure Sass loader in Webpack.
* Added Webpack configuration for production.
* Collapse and expand actions in answers view.
* Client and server questions pagination.
* Realtime answer updates with Websocket.
* GitHub OAuth Login.

### CODE REVIEW

**Comment by yamalight** [comment]
oh, boy. that's a lot of stuff! it'll take some time to go through :)

File: client/src/store/epics/questions.js [hunk 1]
```diff
@@ -12,10 +13,15 @@ export const getAllQuestions = action$ => action$
...
-      type: ActionTypes.GET_ALL_QUESTIONS_ERROR,
-      payload: {error},
-    })),
+    .catch(error => Observable.of(
+      {
+        type: ActionTypes.GET_ALL_QUESTIONS_ERROR,
+        payload: {error},
+      },
+      Actions.addNotificationAction(
+        {text: `[get all questions] Error: ${ajaxErrorToMessage(error)}`, alertType: 'danger'},
```
Comment (yamalight) [inline client/src/store/epics/questions.js]: A bit more human-readable error message would be good.

File: client/src/store/epics/questions.js [hunk 2]
```diff
@@ -24,14 +30,24 @@ export const answerQuestion = action$ => action$
...
+        {text: `Answer: "${payload.answer}" added to question: "${question.text}"`, alertType: 'info'},
+      ),
+    ))
+    .catch(error => Observable.of(
+      {
+        type: ActionTypes.ANSWER_QUESTION_ERROR,
+        payload: {error},
+      },
+      Actions.addNotificationAction(
+        {text: `[answer create] Error: ${ajaxErrorToMessage(error)}`, alertType: 'danger'},
```
Comment (yamalight) [inline client/src/store/epics/questions.js]: Same here.

File: client/src/store/epics/questions.js [hunk 3]
```diff
@@ -40,13 +56,22 @@ export const createQuestion = action$ => action$
...
+        {text: `Question with text "${question.text}" created`, alertType: 'info'},
+      ),
+    ))
+    .catch(error => Observable.of(
+      {
+        type: ActionTypes.CREATE_QUESTION_ERROR,
+        payload: {error},
+      },
+      Actions.addNotificationAction(
+        {text: `[question create] Error: ${ajaxErrorToMessage(error)}`, alertType: 'danger'},
```
Comment (yamalight) [inline client/src/store/epics/questions.js]: And here.

File: client/src/app/index.js [hunk 4]
```diff
@@ -1,10 +1,12 @@
 // npm packages
 import React from 'react';
 
+import Header from '../components/header';
 import Footer from '../components/footer';
 
 export default ({children}) => (
   <div className="container">
+    <Header />
```
Comment (yamalight) [inline client/src/app/index.js]: But this way header will be visible when user is not authenticated. And we don't want that because it makes it confusing for a user.

**Comment by yamalight** [comment]
Ok, this is not going to work well.
I'm gonna close this because it's _really_ hard to review all features mashed together like this - I review change in one of the first commits only to find out it's been outdated by later commit.
Please, create one PR per feature (e.g. one for navbar, one for auth, etc) - will be way easier to work like this.

File: client/package.json [hunk 5]
```diff
@@ -51,6 +51,7 @@
   "dependencies": {
     "bootstrap": "^3.3.7",
     "history": "^4.2.0",
+    "jquery": "^3.1.1",
```
Comment (yamalight) [inline client/package.json]: We are already using React to control DOM. I don't see why we would need jQuery.

File: client/src/index.js [hunk 6]
```diff
@@ -9,6 +9,10 @@ import {Provider} from 'react-redux';
 // styles
 import 'bootstrap/dist/css/bootstrap.min.css';
 
+// JQuery for Bootstrap
+global.jQuery = require('jquery/dist/jquery.min.js');
+require('bootstrap/dist/js/bootstrap.min.js');
```
Comment (yamalight) [inline client/src/index.js]: This should be removed.

File: client/src/components/navbar/index.js [hunk 7]
```diff
@@ -13,17 +13,30 @@ const NavBar = ({actualPath}) => (
   <nav className="navbar navbar-default">
     <div className="container-fluid">
       <div className="navbar-header">
+        <button
+          type="button"
+          className="navbar-toggle collapsed"
+          data-toggle="collapse"
+          data-target="#navbar"
```
Comment (yamalight) [inline client/src/components/navbar/index.js]: Collapse should be done using React, not jQuery.

File: client/src/app/index.js [hunk 8]
```diff
@@ -1,13 +1,43 @@
...
+  actualPath: state.routing.locationBeforeTransitions && state.routing.locationBeforeTransitions.pathname,
+  user: state.auth.user,
+  token: state.auth.token,
+});
+
+const mapDispatchToProps = dispatch => ({
+  navTo: (path) => dispatch(push(path)),
+});
+
+class App extends Component {
```
Comment (yamalight) [inline client/src/app/index.js]: 1. App should be a dumb container without any logic.
2. This might result in an infinite redirect. I think using [redux-auth-wrapper](https://github.com/mjrussell/redux-auth-wrapper) for this would be a better solution.

Comment (yamalight) [inline client/src/app/index.js]: arguably, whole commit can be removed.

File: client/src/pages/login/index.js [hunk 9]
```diff
@@ -26,42 +26,44 @@ const Login = ({onLoginClick}) => {
   };
 
   return (
-    <div className="jumbotron">
-      <h2>Experts portal:</h2>
-      <p>Please log in. Or <Link to="/register">register</Link></p>
+    <div className="container" style={{marginTop: '100px'}}>
```
Comment (yamalight) [inline client/src/pages/login/index.js]: we already have a container in `<App/>`. Why add another one?

File: client/src/pages/register/index.js [hunk 10]
```diff
@@ -26,43 +26,45 @@ const Register = ({onRegisterClick}) => {
   };
 
   return (
-    <div className="jumbotron">
-      <h2>Experts portal:</h2>
-      <p>Please register. Or <Link to="/login">login</Link></p>
+    <div className="container" style={{marginTop: '100px'}}>
```
Comment (yamalight) [inline client/src/pages/register/index.js]: same.

File: client/src/components/footer/index.js [hunk 11]
```diff
@@ -5,18 +5,17 @@ import {Notifications} from '../notifications';
 
 const style = {
   footer: {
-    position: 'absolute',
+    zIndex: 9999,
+    position: 'fixed',
     bottom: '0',
-    width: '100%',
+    width: '90%',
```
Comment (yamalight) [inline client/src/components/footer/index.js]: I'd either use bootstrap styles to scale it properly (e.g. `col-xs-12`) or change styles to use flexbox.

File: client/package.json [hunk 12]
```diff
@@ -58,6 +58,7 @@
     "react-addons-css-transition-group": "^15.4.1",
     "react-dom": "^15.3.2",
     "react-redux": "^4.4.5",
+    "react-responsive": "^1.2.5",
```
Comment (yamalight) [inline client/package.json]: We already use Bootstrap that is responsive by default, why another library?

File: client/package.json [hunk 13]
```diff
@@ -4,7 +4,9 @@
   "description": "Experts portal client",
   "main": "index.js",
   "scripts": {
-    "start": "node serve.js"
+    "start": "node serve.js",
+    "dist": "rm -rf dist/ && webpack -p --config webpack.production.config.js",
```
Comment (yamalight) [inline client/package.json]: Don't you have `dist/redirect.html` there? You killing it every time on build then?

File: client/package.json [hunk 14]
```diff
@@ -49,11 +51,13 @@
     "webpack-hot-middleware": "^2.12.2"
   },
   "dependencies": {
+    "babel-runtime": "^6.18.0",
```
Comment (yamalight) [inline client/package.json]: Why exactly do we need it?

File: client/package.json [hunk 15]
```diff
@@ -49,11 +51,13 @@
...
     "bootstrap": "^3.3.7",
     "history": "^4.2.0",
     "jquery": "^3.1.1",
     "lodash": "^4.17.2",
     "moment": "^2.17.0",
+    "pushstate-server": "^2.1.0",
```
Comment (yamalight) [inline client/package.json]: And this one too?

File: client/webpack.production.config.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
+const path = require('path');
```
Comment (yamalight) [inline client/webpack.production.config.js]: I don't see enough differences with test config to justify another file.

File: client/src/store/index.js [hunk 17]
```diff
@@ -20,6 +20,14 @@ const middlewares = composeEnhancers(
 );
 
 // create store
-const store = createStore(rootReducer, middlewares);
+const store = createStore(rootReducer, {}, middlewares);
```
Comment (yamalight) [inline client/src/store/index.js]: Preloaded state is optional. Any reason to include it explicitly?

**Comment by yamalight** [comment]
@surtich ok, I think I'd gone through most of the code. you can start replying/reviewing now :)

File: client/src/store/epics/questions.js (hunk 3 above)
Comment (surtich) [inline client/src/store/epics/questions.js]: I have improved the legibility of these three. But I do not know what you mean with this: "Why is this necessary? Can't we just use devtools request throttling for this?"

File: server/config.js [hunk 18]
```diff
@@ -11,4 +11,10 @@ exports.auth = {
     'RGP84d%XZ$tck7TPpQ^zn#7Q$i&duxS2K!8ZR!87!9vJ2yZe@ZFqSMIvdvv4EseS',
   jwtSecret: process.env.EXPERTS_AUTH_JWTSECRET ||
     'uaeldt!2D9iVrOv1KEH#KRuaiEdJty6rRXJij$FN&D$oYKITos14Utok6W0kt83@',
+  github: {
+    clientID: process.env.GITHUB_CLIENT_ID || 'b31b524fd5e239d5dc68',
```
Comment (yamalight) [inline server/config.js]: are those real btw? if yes - you better change them in your app 😆

File: client/src/components/navbar/index.js (hunk 7 above)
Comment (surtich) [inline client/src/components/navbar/index.js]: Done the last three. I've been lazy with this. I have not been able to get the animation of the NavBar. I leave it as a pending task. Maybe you could use [react-collapse](https://github.com/nkbt/react-collapse).

File: client/src/app/index.js (hunk 8 above)
Comment (surtich) [inline client/src/app/index.js]: You're right, this is not a very good solution but I think it's better than previously. I think we should definitely use redux-auth-wrapper. Now I do not have time but I can investigate it later.

File: client/src/components/footer/index.js (hunk 11 above)
Comment (surtich) [inline client/src/components/footer/index.js]: Done it.

File: client/package.json (hunk 12 above)
Comment (yamalight) [inline client/package.json]: @surtich I'd say there should be only `<QuestionList />` there and the component itself should be responsive.

File: client/src/app/index.js (hunk 8 above)
Comment (yamalight) [inline client/src/app/index.js]: @surtich then please just remove those commits, I'll add auth-wrapper myself :)

File: client/src/store/epics/questions.js (hunk 3 above)
Comment (yamalight) [inline client/src/store/epics/questions.js]: @surtich "Why is this necessary?" was related to commit with RxJS delay stuff.

**Comment by surtich** [comment]
I think I explained the reason why I had to do such a great PR. I close the PR and try to contribute in the future following your indications. Thanks.

**Comment by yamalight** [comment]
@surtich as I said in the video - I just used this PR as an example. 
I think we can still successfully merge it if you sort out all my comments :)
But if you are willing to make smaller ones - I'll be only happy :D
